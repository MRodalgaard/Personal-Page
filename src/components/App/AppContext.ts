import { createContext } from 'react';
import { AppBackground } from '../../util/constants';
import { AppColors } from '../../util/theme';

export interface IAppState {
  color: AppColors;
  background: AppBackground;
  toggleBackground?: () => void;
  toggleColor?: () => void;
}

export const initialState: IAppState = {
  background: AppBackground.greyscale,
  color: AppColors.primary,
};

const AppContext = createContext(initialState);

export default AppContext;
