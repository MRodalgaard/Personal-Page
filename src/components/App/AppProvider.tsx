import React, { ReactNode, useReducer } from 'react';
import createPersistedReducer from 'use-persisted-reducer';
import { ActionType } from '../../store/actions';
import reducer from '../../store/reducer';
import { AppBackground, LOCALSTORAGE_KEY } from '../../util/constants';
import { AppColors } from '../../util/theme';
import AppContext, { initialState } from './AppContext';

const usePersistedReducer: typeof useReducer = createPersistedReducer(
  LOCALSTORAGE_KEY
);

interface IProps {
  children: ReactNode;
}

const AppProvider = ({ children }: IProps) => {
  // Persist state to localstorage after reducer has been called
  const [state, dispatch] = usePersistedReducer(reducer, initialState);

  // Inject dispatch actions into provider value
  const value = {
    ...state,
    toggleBackground: () => {
      dispatch({
        payload:
          state.background === AppBackground.colored
            ? AppBackground.greyscale
            : AppBackground.colored,
        type: ActionType.UPDATE_BACKGROUND,
      });
    },
    toggleColor: () => {
      dispatch({
        payload:
          state.color === AppColors.primary
            ? AppColors.secondary
            : AppColors.primary,
        type: ActionType.UPDATE_COLOR,
      });
    },
  };

  return <AppContext.Provider value={value}>{children}</AppContext.Provider>;
};

export default AppProvider;
