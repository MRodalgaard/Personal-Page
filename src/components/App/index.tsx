import * as React from 'react';
import styled from 'styled-components';
import Background from '../Background';
import Envelope from '../Envelope';
import Letter from '../Letter';
import AppProvider from './AppProvider';

const Content = styled.div`
  margin: 20px;

  @media (min-width: 500px) {
    float: right;
    width: 400px;
  }
`;

const App = () => {
  return (
    <AppProvider>
      <Background />
      <Content>
        <Envelope />
        <Letter />
      </Content>
    </AppProvider>
  );
};

export default App;
