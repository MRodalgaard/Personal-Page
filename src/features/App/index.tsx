import { Age } from 'components/Age';
import { Background } from 'components/Background';
import { Envelope, EnvelopeMiddle, EnvelopeTop } from 'components/Envelope';
import { Icon } from 'components/Icon';
import { Letter } from 'components/Letter';
import { Link } from 'components/Link';
import { MailLink } from 'components/MailLink';
import { ModeButton } from 'components/ModeButton';
import { Quote } from 'components/Quote';
import { RightSideContent } from 'components/RightSideContent';
import { SoundButton } from 'components/SoundButton';
import { Spacer } from 'components/Spacer';
import { useAppContext } from 'contexts/AppContext';
import { AnalyticsEvent } from 'utils/analytics';

export const App = () => {
  const { toggleColorized } = useAppContext();

  return (
    <>
      <Background />
      <RightSideContent>
        <Envelope>
          <EnvelopeTop>
            <Link
              href="https://linkedin.com/in/mrodalgaard"
              analyticsEvent={AnalyticsEvent.LinkedInIconLink}
              ariaLabel="LinkedIn"
            >
              <Icon type="linkedIn" />
            </Link>
            <Link href="https://x.com/mrodalgaard" analyticsEvent={AnalyticsEvent.XIconLink} ariaLabel="X">
              <Icon type="x" />
            </Link>
            <Link
              href="https://github.com/mrodalgaard"
              analyticsEvent={AnalyticsEvent.GithubIconLink}
              ariaLabel="GitHub"
            >
              <Icon type="github" />
            </Link>
            <Spacer />
            <ModeButton />
            <SoundButton />
            <MailLink href="mailto:mrodalgaard@gmail.com" />
          </EnvelopeTop>

          <EnvelopeMiddle>
            Martin Rodalgaard
            <br />
            Aarhus
            <br />
            Denmark
          </EnvelopeMiddle>
        </Envelope>
        <Letter>
          <p>
            Dear{' '}
            <Link href="https://whatsmybrowser.org/" analyticsEvent={AnalyticsEvent.WhoAmILink}>
              Reader
            </Link>
            ,
          </p>
          <p>
            <Link href="https://x.com/mrodalgaard" analyticsEvent={AnalyticsEvent.XLink}>
              Martin Rodalgaard St√∏rup
            </Link>{' '}
            is <Age birthday="1986-08-22">&nbsp;&nbsp;</Age> years and living in Aarhus, Denmark, with his lovely wife
            Rikke and wonderful children Gry &amp; Thor.
          </p>
          <p>
            He has an education in{' '}
            <Link href="https://linkedin.com/in/mrodalgaard" analyticsEvent={AnalyticsEvent.LinkedInLink}>
              Electrical Engineering
            </Link>{' '}
            specialised in Embedded Systems and{' '}
            <Link href="https://github.com/mrodalgaard" analyticsEvent={AnalyticsEvent.GithubLink}>
              Programming
            </Link>
            . He currently works at{' '}
            <Link href="https://trifork.com" analyticsEvent={AnalyticsEvent.WorkLink}>
              Trifork A/S
            </Link>{' '}
            creating exciting mobile and web app solutions for customers.
          </p>
          <p>
            Martin loves{' '}
            <Link onClick={toggleColorized} analyticsEvent={AnalyticsEvent.ColorClick}>
              colors
            </Link>
            , big city traveling, running, skiing, watching movie classics and reading biographies.
          </p>
          <p>
            The gear he <i>can</i> live without (but doesn&#39;t want to) includes his 16&quot; MacBook Pro, iPhone, an
            old Canon camera, a converted Giant racing bike and an Omega Seamaster on his wrist.
          </p>

          <Quote />
        </Letter>
      </RightSideContent>
    </>
  );
};
