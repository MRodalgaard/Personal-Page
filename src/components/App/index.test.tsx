import { shallow } from "enzyme";
import * as React from "react";
import App from ".";

describe("App", () => {
  it("renders", () => {
    const wrapper = shallow(<App />);
    expect(wrapper).toMatchSnapshot();
  });
});
