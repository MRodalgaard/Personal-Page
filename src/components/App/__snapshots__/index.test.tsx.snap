// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders 1`] = `
<AppProvider>
  <Background />
  <styled.div>
    <Envelope />
    <Letter />
  </styled.div>
</AppProvider>
`;
